sstables/sstable_set.cc

```diff
@@ -464,6 +470,105 @@ class time_series_sstable_set : public sstable_set_impl {
+    container_t::const_iterator _it;
+    const container_t::const_iterator _end;
+
+    position_in_partition::tri_compare _cmp;
+
+    std::function<flat_mutation_reader(sstable&)> _create_reader;
+    std::function<bool(const sstable&)> _filter;
+
+    flat_mutation_reader create_reader(sstable& sst) {
+        return _create_reader(sst);
+    }
+
+    bool filter(const sstable& sst) const {
+        return _filter(sst);
+    }
+
+public:
+    min_position_reader_queue(schema_ptr schema,
+            lw_shared_ptr<const time_series_sstable_set::container_t> sstables,
+            std::function<flat_mutation_reader(sstable&)> create_reader,
+            std::function<bool(const sstable&)> filter)
+        : _schema(std::move(schema))
+        , _sstables(std::move(sstables))
+        , _it(_sstables->begin())
+        , _end(_sstables->end())
+        , _cmp(*_schema)
+        , _create_reader(std::move(create_reader))
+        , _filter(std::move(filter)) {}
+
+    // Open sstable readers to all sstables with smallest min_position() from the set
+    // {S: filter(S) and prev_min_pos < S.min_position() <= bound}, where `prev_min_pos` is the min_position()
+    // of the sstables returned from last non-empty pop() or -infinity if no sstables were previously returned,
+    // and `filter` is the filtering function provided when creating the queue.
+    //
+    // Note that there may be multiple returned sstables (all with the same position) or none.
+    //
+    // Note that S.min_position() is global for sstable S; if the readers are used to inspect specific partitions,
+    // the minimal positions in these partitions might actually all be greater than S.min_position().
+    std::vector<reader_and_positions> pop(position_in_partition_view bound, db::timeout_clock::time_point timeout) {
+        if (empty(bound)) {
+            return {};
+        }
+
+
+        auto it = std::find_if(_it, _end, [this, &bound] (const value_t& v) { return filter(*v.second); });
+        assert(_cmp(it->first, bound) < 0); // because (!empty(bound))
+
+        // Find all sstables with the same position as `it` (they form a contiguous range in the container).
+        auto next = std::find_if(it, _end, [this, it] (const value_t& v) { return _cmp(v.first, it->first) != 0; });
+
+        std::vector<reader_and_positions> ret;
+        ret.reserve(std::distance(it, next));
+        for (; it != next; ++it) {
+            if (filter(*it->second)) {
+                ret.emplace_back(create_reader(*it->second), it->second->max_position());
+            }
+        }
+
+        _it = next;
+        return ret;
+    }
+
+    // Is the set of sstables {S: filter(S) and prev_min_pos < S.min_position() <= bound} empty?
+    // (see pop() for definition of `prev_min_pos`)
+    bool empty(const position_in_partition_view bound) const {
+        auto past_bound = std::find_if(_it, _end, [this, bound] (const value_t& v) { return _cmp(bound, v.first) < 0; });
```

Comment:
`std::lower_bound`.

Reply:
`upper_bound` actually.
But does it matter? I find `find_if` (no pun intended) more readable (it's just conceptually simpler), and using `upper_bound` won't speed things up since we're dealing with a non-random iterator here.

Reply:
I think `lower_bound`/`upper_bound` is more expressive of the intent at glance, while with `std::find_if`, one has to parse the entire expression to understand it.